kernel/src/log_segment.rs

```diff
@@ -57,126 +70,93 @@ pub(crate) struct LogSegment {
-                    "LogSegment end version {effective_version} not the same as the specified end version {end_version}"
-                ))
-            );
+impl LogSegmentBuilder {
+    fn new(log_root: Url) -> Self {
+        LogSegmentBuilder {
+            log_root,
+            end_version: None,
         }
+    }
 
-        Ok(LogSegment {
-            end_version: effective_version,
-            checkpoint_version,
-            log_root,
-            ascending_commit_files,
-            ascending_compaction_files,
-            checkpoint_parts,
-            latest_crc_file,
-        })
+    /// Sets the end version for the log segment. This is optional, and if not set, the LogSegment
+    /// will LIST to determine the end version at the latest version of the table.
+    #[internal_api]
+    pub(crate) fn with_end_version_opt(mut self, end_version: Option<Version>) -> Self {
+        self.end_version = end_version;
+        self
     }
 
-    /// Constructs a [`LogSegment`] to be used for [`Snapshot`]. For a `Snapshot` at version `n`:
-    /// Its LogSegment is made of zero or one checkpoint, and all commits between the checkpoint up
-    /// to and including the end version `n`. Note that a checkpoint may be made of multiple
-    /// parts. All these parts will have the same checkpoint version.
+    /// Constructs a complete [`LogSegment`]. Complete is defined as a `LogSegment` which includes
+    /// _all_ actions on a table up to and including the end version. These `LogSegment`s are used
+    /// for [`Snapshot`]s, [`Scan`]s, etc.
     ///
-    /// The options for constructing a LogSegment for Snapshot are as follows:
-    /// - `checkpoint_hint`: a `LastCheckpointHint` to start the log segment from (e.g. from reading the `last_checkpoint` file).
-    /// - `time_travel_version`: The version of the log that the Snapshot will be at.
+    /// Specifically, for a complete `Snapshot` at version `n`:
+    /// Its `LogSegment` is made of zero or one checkpoint, and all commits between the checkpoint
+    /// up to and including the end version `n`. Note that a checkpoint may be made of multiple
+    /// parts. All these parts will have the same checkpoint version. Also note the commits may be
+    /// included as individual commit files or as part of log-compaction files.
     ///
     /// [`Snapshot`]: crate::snapshot::Snapshot
+    /// [`Scan`]: crate::scan::Scan
     #[internal_api]
-    pub(crate) fn for_snapshot(
-        storage: &dyn StorageHandler,
-        log_root: Url,
-        checkpoint_hint: impl Into<Option<LastCheckpointHint>>,
-        time_travel_version: impl Into<Option<Version>>,
-    ) -> DeltaResult<Self> {
-        let time_travel_version = time_travel_version.into();
+    pub(crate) fn build_complete(self, storage: &dyn StorageHandler) -> DeltaResult<LogSegment> {
+        let checkpoint_hint = LastCheckpointHint::try_read(storage, &self.log_root)?;
+        self.build_complete_impl(storage, checkpoint_hint)
```

Comment:
I wonder if we really need to remove the checkpoint hint?

The build call is now implicitly doing 2 network calls: one to read the checkpoint hint, and another to do file listing. But what if the user wants to choose not to read checkpoint hint, or they want to cache it and reuse it for multiple snapshot calls?